src/include/util.coffee

```diff
@@ -22,4 +22,8 @@ balanced = (name, left, right, inner, ignore = '') ->
   else
     "(?<#{name}>(?:[^#{left}#{right}#{ignore}]|#{left}\\g<#{name}>#{right})*)"
 
-module.exports = {list, listMaybe, concat, balanced}
+floatPattern = (digit, exp) ->
+  "#{digit}*(?:(?:\\.#{digit}+)?#{exp}[+-]?[0-9_]+|\\.#{digit}+)"
```

Comment:
Okay, now this expression here doesn't match floats without explicit exponents, e.g. `3.1415`.

In general, Haskell floating points have syntactical from similar to one of:
- `3.1415`
- `314e-2`
- `31.415e-1`

I would imagine it's similar for hexfloats?

Anyway, we'd like to catch all of those, that's why original float regular expression is kinda convoluted -- you can have either fractional part, exponent part, or both of them, but you can't have a float without at least one. In quasi-EBNF terms, this would be somewhat simple to express:
```bnf
float → digit+ ( fraction | exponent | fraction exponent )
fraction → "." digit+
exponent → [eE] [+-]? digit+
```
Equivalent regular expression was trivial to factorize, because all alternatives end on `digit+`, but if `digit` in `exponent` can differ from `digit` elsewhere, this won't work, so you'll need to write out alternatives.

However, I should point out, `fraction` and `fraction exponent` have the same prefix, and hence would require backtracking. We can avoid that by rewriting expression a little:
```bnf
float → digit+ ( fraction exponent? | exponent )
```
this way, regexp engine can bail from any branch early

Reply:
Okay, disregard the first line, I made a mistake when testing. Still, not starting branches on optional parts will on average be about twice as effective.

Reply:
> which you apparently have here (notice common prefix, remember that regexp parses left-to-right, and realize that fraction part will be effectively matched twice if there is no exponent)

Thank you for teaching me about effective regexp.
You gave me a wonderful insight about behavior of parsing regex.

I will revise the pattern.